src/realm/sync/client.cpp

```diff
@@ -1457,18 +1509,24 @@ bool SessionWrapper::wait_for_download_complete_or_client_stopped()
 void SessionWrapper::refresh(std::string signed_access_token)
 {
     // Thread safety required
-    REALM_ASSERT(m_initiated);
-    REALM_ASSERT(!m_finalized);
+    {
```

Comment:
This is essentially very suspicious code when you guard just a few variables with very limited scope. Looks like this class serves multiple purposes... I'd need to look at it more carefully. It's not obvious it this fix just silence particular tsan warning or solves the issue with this class design and concurrent access.